DEFI/LiquidityPool/certora/specs/Full.spec

```diff
@@ -0,0 +1,448 @@
+}
+
+rule depositRevertConditions(env e){
+
+    require e.msg.sender != currentContract; // this assumption must hold to avoid shares dilute attack
+
+    uint256 amount;
+    uint256 clientBalanceBefore = underlying.balanceOf(e.msg.sender);
+    uint256 clientSharesBefore = balanceOf(e.msg.sender);
+    
+    bool underFlow = clientBalanceBefore - amount < 0;
+    bool emptyPool = totalSupply() == 0 || underlying.balanceOf(currentContract) == 0;
+    bool clientSharesOverflow = (clientSharesBefore + amount > max_uint256 && emptyPool) || clientSharesBefore + amountToShares(amount) > max_uint256;
+    bool totalSharesOverflow = totalSupply() + amountToShares(amount) > max_uint256;
+    bool contractUnderlyingOverflow = underlying.balanceOf(currentContract) + amount > max_uint256;
+    bool overflow =  clientSharesOverflow || totalSharesOverflow || contractUnderlyingOverflow;
+    bool payable = e.msg.value != 0;
+    bool reentrancy = ghostReentrancyStatus == 2;
+    bool notEnoughAllowance = underlying.allowance(e.msg.sender, currentContract) < amount;
+    bool zeroShares = amountToShares(amount) == 0 && totalSupply() > 0 && underlying.balanceOf(currentContract) > 0;
+    bool expectedRevert = underFlow || overflow || payable || reentrancy || notEnoughAllowance || zeroShares;
+
+    deposit@withrevert(e, amount);
+
+    assert lastReverted <=> expectedRevert;
+}
+
+rule depositDontAffectThirdParty(env e){
+    address thirdParty;
+    uint256 amount;
+
+    require thirdParty != e.msg.sender && thirdParty != currentContract;
+
+    uint256 thirdPartyBalanceBefore = underlying.balanceOf(thirdParty);
+    uint256 thirdPartySharesBefore = balanceOf(thirdParty);
+
+    deposit(e, amount);
+
+    uint256 thirdPartyBalanceAfter = underlying.balanceOf(thirdParty);
+    uint256 thirdPartySharesAfter = balanceOf(thirdParty);
+
+    assert thirdPartyBalanceBefore == thirdPartyBalanceAfter;
+    assert thirdPartySharesBefore == thirdPartySharesAfter;
+}
+
+rule depositGreaterThanZeroWithMinted(env e) {
+    uint256 amount;
+    uint256 amountMinted = deposit(e, amount);
+    
+    assert amount > 0 <=> amountMinted > 0;
+}
+
+rule splitDepositFavoursTheContract(env e){
+    require e.msg.sender != currentContract; // this assumption must hold to avoid shares dilute attack
+
+    uint256 wholeAmount;
+    uint256 amountA; 
+    uint256 amountB;
+    require amountA + amountB == to_mathint(wholeAmount);
+
+    requireInvariant totalSharesLessThanUnderlyingBalance();
+    requireInvariant totalIsSumBalances();
+    requireInvariant totalSharesEqualSumOfShares();
+    requireInvariant totalSharesIsZeroWithUnderlying();
+
+    storage init = lastStorage;
+
+    uint256 wholeShares = deposit(e, wholeAmount);
+
+    uint256 sharesA = deposit(e, amountA) at init;
+    uint256 sharesB = deposit(e, amountB);
+
+    assert to_mathint(wholeShares) >= sharesA + sharesB;
+}
```

Comment:
did we prove that the return value of deposit is what added to the balancOf? I think this is missing and can be added to integrity

Reply:
Done, also for withdraw